src/Microsoft.Data/GroupBy.cs

```diff
@@ -0,0 +1,432 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Data
+{
+    /// <summary>
+    /// A GroupBy class that is typically the product of a DataFrame.GroupBy call
+    /// It holds information to perform typical aggregation ops on it
+    /// </summary>
+    public abstract class GroupBy
+    {
+        /// <summary>
+        /// Returns true if all values in the group are true
+        /// </summary>
+        /// <returns></returns>
+        public abstract DataFrame All();
+
+        /// <summary>
+        /// Compute the number of values in each group 
+        /// </summary>
+        /// <returns></returns>
+        public abstract DataFrame Count();
+
+        /// <summary>
+        /// Return the first value in each group
+        /// </summary>
+        /// <returns></returns>
+        public abstract DataFrame First();
+
+        /// <summary>
+        /// Returns the first numberOfRows rows of each group
+        /// </summary>
+        /// <param name="numberOfRowsInEachGroup"></param>
+        /// <returns></returns>
+        public abstract DataFrame Head(int numberOfRows);
+
+        /// <summary>
+        /// Returns the last numberOfRows rows of each group
+        /// </summary>
+        /// <param name="numberOfRowsInEachGroup"></param>
+        /// <returns></returns>
+        public abstract DataFrame Tail(int numberOfRows);
+
+        /// <summary>
+        /// Compute the max of group values
+        /// </summary>
+        /// <returns></returns>
+        public abstract DataFrame Max();
+        public abstract DataFrame Min();
+        public abstract DataFrame Product();
+        public abstract DataFrame Sum();
+    }
+
+    public class GroupBy<TKey> : GroupBy
+    {
+        private int _groupByColumnIndex;
+        private Dictionary<TKey, ICollection<long>> _keyValuePairs;
```

Comment:
`_keyValuePairs` doesn't seem like the best name for this variable. Does `_keysToRowIndices` describe it better?